base/cvd/cuttlefish/host/commands/host_bugreport/main.cc

```diff
@@ -31,30 +33,34 @@
 #include "common/libs/utils/subprocess.h"
 #include "common/libs/utils/tee_logging.h"
 #include "host/libs/config/cuttlefish_config.h"
-#include "ziparchive/zip_writer.h"
 
 DEFINE_string(output, "host_bugreport.zip", "Where to write the output");
 DEFINE_bool(include_adb_bugreport, false, "Includes device's `adb bugreport`.");
 
 namespace cuttlefish {
 namespace {
 
-void SaveFile(ZipWriter& writer, const std::string& zip_path,
+void SaveFile(zip_t* archive, const std::string& zip_path,
               const std::string& file_path) {
-  writer.StartEntry(zip_path, ZipWriter::kCompress | ZipWriter::kAlign32);
-  std::fstream file(file_path, std::fstream::in | std::fstream::binary);
-  do {
-    char data[1024 * 10];
-    file.read(data, sizeof(data));
-    writer.WriteBytes(data, file.gcount());
-  } while (file);
-  writer.FinishEntry();
-  if (file.bad()) {
-    LOG(ERROR) << "Error in logging " << file_path << " to " << zip_path;
+  if (!FileExists(file_path)) {
+    LOG(ERROR) << "Tried to add file '" << file_path << "', does not exist.";
+    return;
+  }
+  zip_source_t* source =
+      zip_source_file(archive, file_path.c_str(), 0, ZIP_LENGTH_TO_END);
+  if (source == nullptr) {
+    LOG(ERROR) << "Error in creating zip source file: "
+               << zip_strerror(archive);
+    return;
+  }
+  zip_int64_t index = zip_file_add(archive, zip_path.c_str(), source, 0);
+  if (index == -1) {
+    LOG(ERROR) << "Error in adding zip source file: " << zip_strerror(archive);
+    zip_source_free(source);
```

Comment:
should this free be outside the `if`? we should consider a RAII wrapper around this interface too and hide this raw pointer handling in a library somewhere.

Reply:
Yes. `zip_file_add` [consumes the `zip_source_t*` on success](https://libzip.org/documentation/zip_file_add.html).

> _NOTE_: [`zip_source_free(3)`](https://libzip.org/documentation/zip_source_free.html) should not be called on a source after it was used successfully in a `zip_file_add` or `zip_file_replace` call.